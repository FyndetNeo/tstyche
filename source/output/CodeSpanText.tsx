import type { TestMember, TestTree } from "#collect";
import type { DiagnosticOrigin } from "#diagnostic";
import { Path } from "#path";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";

interface BreadcrumbsTextProps {
  ancestor: TestMember | TestTree;
}

function BreadcrumbsText({ ancestor }: BreadcrumbsTextProps) {
  const text: Array<string> = [];

  while ("name" in ancestor) {
    text.push(ancestor.name);
    ancestor = ancestor.parent;
  }

  text.push("");

  return <Text color={Color.Gray}>{text.reverse().join(" ‚ù≠ ")}</Text>;
}

interface CodeLineTextProps {
  gutterWidth: number;
  lineNumber: number;
  lineNumberColor?: Color;
  lineText: string;
}

function CodeLineText({ gutterWidth, lineNumber, lineNumberColor = Color.Gray, lineText }: CodeLineTextProps) {
  return (
    <Line>
      <Text color={lineNumberColor}>{lineNumber.toString().padStart(gutterWidth)}</Text>
      <Text color={Color.Gray}>{" | "}</Text>
      {lineText}
    </Line>
  );
}

interface SquiggleLineTextProps {
  gutterWidth: number;
  indentWidth?: number;
  squiggleWidth: number;
}

function SquiggleLineText({ gutterWidth, indentWidth = 0, squiggleWidth }: SquiggleLineTextProps) {
  return (
    <Line>
      {" ".repeat(gutterWidth)}
      <Text color={Color.Gray}>{" | "}</Text>
      {" ".repeat(indentWidth)}
      <Text color={Color.Red}>{"~".repeat(squiggleWidth === 0 ? 1 : squiggleWidth)}</Text>
    </Line>
  );
}

interface CodeSpanTextProps {
  diagnosticOrigin: DiagnosticOrigin;
}

export function CodeSpanText({ diagnosticOrigin }: CodeSpanTextProps) {
  const lastLineInFile = diagnosticOrigin.sourceFile.getLineAndCharacterOfPosition(
    diagnosticOrigin.sourceFile.text.length,
  ).line;

  const { character: firstMarkedLineCharacter, line: firstMarkedLine } =
    diagnosticOrigin.sourceFile.getLineAndCharacterOfPosition(diagnosticOrigin.start);
  const { character: lastMarkedLineCharacter, line: lastMarkedLine } =
    diagnosticOrigin.sourceFile.getLineAndCharacterOfPosition(diagnosticOrigin.end);

  const firstLine = Math.max(firstMarkedLine - 2, 0);
  const lastLine = Math.min(firstLine + 5, lastLineInFile);
  const gutterWidth = (lastLine + 1).toString().length + 2;

  const codeSpan: Array<ScribblerJsx.Element> = [];

  for (let index = firstLine; index <= lastLine; index++) {
    const lineStart = diagnosticOrigin.sourceFile.getPositionOfLineAndCharacter(index, 0);
    const lineEnd =
      index === lastLineInFile
        ? diagnosticOrigin.sourceFile.text.length
        : diagnosticOrigin.sourceFile.getPositionOfLineAndCharacter(index + 1, 0);

    const lineText = diagnosticOrigin.sourceFile.text.slice(lineStart, lineEnd).trimEnd().replace(/\t/g, " ");

    if (index >= firstMarkedLine && index <= lastMarkedLine) {
      codeSpan.push(
        <CodeLineText
          gutterWidth={gutterWidth}
          lineNumber={index + 1}
          lineNumberColor={Color.Red}
          lineText={lineText}
        />,
      );

      if (index === firstMarkedLine) {
        const squiggleLength =
          index === lastMarkedLine
            ? lastMarkedLineCharacter - firstMarkedLineCharacter
            : lineText.length - firstMarkedLineCharacter;

        codeSpan.push(
          <SquiggleLineText
            gutterWidth={gutterWidth}
            indentWidth={firstMarkedLineCharacter}
            squiggleWidth={squiggleLength}
          />,
        );
      } else if (index === lastMarkedLine) {
        codeSpan.push(<SquiggleLineText gutterWidth={gutterWidth} squiggleWidth={lastMarkedLineCharacter} />);
      } else {
        codeSpan.push(<SquiggleLineText gutterWidth={gutterWidth} squiggleWidth={lineText.length} />);
      }
    } else {
      codeSpan.push(<CodeLineText gutterWidth={gutterWidth} lineNumber={index + 1} lineText={lineText} />);
    }
  }

  const location = (
    <Line>
      {" ".repeat(gutterWidth + 2)}
      <Text color={Color.Gray}>{" at "}</Text>
      <Text color={Color.Cyan}>{Path.relative("", diagnosticOrigin.sourceFile.fileName)}</Text>
      <Text color={Color.Gray}>{`:${firstMarkedLine + 1}:${firstMarkedLineCharacter + 1}`}</Text>
      {diagnosticOrigin.assertion && <BreadcrumbsText ancestor={diagnosticOrigin.assertion.parent} />}
    </Line>
  );

  return (
    <Text>
      {codeSpan}
      <Line />
      {location}
    </Text>
  );
}
