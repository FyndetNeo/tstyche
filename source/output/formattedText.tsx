import { Line, type ScribblerJsx } from "#scribbler";

export class JsonText implements ScribblerJsx.ElementClass {
  constructor(readonly props: { input: Array<string> | Record<string, unknown> }) {}

  render(): ScribblerJsx.Element {
    return <Line>{JSON.stringify(this.#sortObject(this.props.input), null, 2)}</Line>;
  }

  #sortObject(target: Array<string> | Record<string, unknown>) {
    if (Array.isArray(target)) {
      return target;
    }

    return Object.keys(target)
      .sort()
      .reduce<Record<string, unknown>>((result, key) => {
        result[key] = target[key];

        return result;
      }, {});
  }
}

export function formattedText(input: string | Array<string> | Record<string, unknown>): ScribblerJsx.Element {
  if (typeof input === "string") {
    return <Line>{input}</Line>;
  }

  return <JsonText input={input} />;
}
