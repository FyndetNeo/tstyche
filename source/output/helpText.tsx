import { OptionBrand, type OptionDefinition } from "#config";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";

const usageExamples: Array<[commandText: string, descriptionText: string]> = [
  ["tstyche", "Run all tests."],
  ["tstyche path/to/first.test.ts", "Only run the test files with matching path."],
  ["tstyche --target 4.9,5.3.2,current", "Test on all specified versions of TypeScript."],
];

interface HintTextProps {
  children: ScribblerJsx.Element;
}

class HintText implements ScribblerJsx.ElementClass {
  props: HintTextProps;

  constructor(props: HintTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    return (
      <Text indent={1} color={Color.Gray}>
        {this.props.children}
      </Text>
    );
  }
}

interface HelpHeaderTextProps {
  tstycheVersion: string;
}

class HelpHeaderText implements ScribblerJsx.ElementClass {
  props: HelpHeaderTextProps;

  constructor(props: HelpHeaderTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    const hint = (
      <HintText>
        <Text>{this.props.tstycheVersion}</Text>
      </HintText>
    );

    return (
      <Line>
        <Text>The TSTyche Type Test Runner</Text>
        {hint}
      </Line>
    );
  }
}

interface CommandTextProps {
  hint?: ScribblerJsx.Element | undefined;
  text: string | ScribblerJsx.Element;
}

class CommandText implements ScribblerJsx.ElementClass {
  props: CommandTextProps;

  constructor(props: CommandTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    let hint: ScribblerJsx.Element | undefined;

    if (this.props.hint != null) {
      hint = <HintText>{this.props.hint}</HintText>;
    }

    return (
      <Line indent={1}>
        <Text color={Color.Blue}>{this.props.text}</Text>
        {hint}
      </Line>
    );
  }
}

interface OptionDescriptionTextProps {
  text: string;
}

class OptionDescriptionText implements ScribblerJsx.ElementClass {
  props: OptionDescriptionTextProps;

  constructor(props: OptionDescriptionTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    return <Line indent={1}>{this.props.text}</Line>;
  }
}

class CommandLineUsageText implements ScribblerJsx.ElementClass {
  render(): ScribblerJsx.Element {
    const usageText = usageExamples.map(([commandText, descriptionText]) => (
      <Text>
        <CommandText text={commandText} />
        <OptionDescriptionText text={descriptionText} />
        <Line />
      </Text>
    ));

    return <Text>{usageText}</Text>;
  }
}

interface CommandLineOptionNameTextProps {
  text: string;
}

class CommandLineOptionNameText implements ScribblerJsx.ElementClass {
  props: CommandLineOptionNameTextProps;

  constructor(props: CommandLineOptionNameTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    return <Text>--{this.props.text}</Text>;
  }
}

interface CommandLineOptionHintTextProps {
  definition: OptionDefinition;
}

class CommandLineOptionHintText implements ScribblerJsx.ElementClass {
  props: CommandLineOptionHintTextProps;

  constructor(props: CommandLineOptionHintTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    if (this.props.definition.brand === OptionBrand.List) {
      return (
        <Text>
          {this.props.definition.brand} of {this.props.definition.items.brand}s
        </Text>
      );
    }

    return <Text>{this.props.definition.brand}</Text>;
  }
}

interface CommandLineOptionsTextProps {
  optionDefinitions: Map<string, OptionDefinition>;
}

class CommandLineOptionsText implements ScribblerJsx.ElementClass {
  props: CommandLineOptionsTextProps;

  constructor(props: CommandLineOptionsTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    const definitions = [...this.props.optionDefinitions.values()];
    const optionsText = definitions.map((definition) => {
      let hint: ScribblerJsx.Element | undefined;

      if (definition.brand !== OptionBrand.BareTrue) {
        hint = <CommandLineOptionHintText definition={definition} />;
      }

      return (
        <Text>
          <CommandText text={<CommandLineOptionNameText text={definition.name} />} hint={hint} />
          <OptionDescriptionText text={definition.description} />
          <Line />
        </Text>
      );
    });

    return (
      <Text>
        <Line>Command Line Options</Line>
        <Line />
        {optionsText}
      </Text>
    );
  }
}

class HelpFooterText implements ScribblerJsx.ElementClass {
  render(): ScribblerJsx.Element {
    return <Line>To learn more, visit https://tstyche.org</Line>;
  }
}

export function helpText(
  optionDefinitions: Map<string, OptionDefinition>,
  tstycheVersion: string,
): ScribblerJsx.Element {
  return (
    <Text>
      <HelpHeaderText tstycheVersion={tstycheVersion} />
      <Line />
      <CommandLineUsageText />
      <Line />
      <CommandLineOptionsText optionDefinitions={optionDefinitions} />
      <Line />
      <HelpFooterText />
      <Line />
    </Text>
  );
}
