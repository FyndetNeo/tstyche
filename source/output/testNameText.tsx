import { Color, Line, type ScribblerJsx, Text } from "#scribbler";

interface StatusTextProps {
  status: "fail" | "pass" | "skip" | "todo";
}

class StatusText implements ScribblerJsx.ElementClass {
  props: StatusTextProps;

  constructor(props: StatusTextProps) {
    this.props = props;
  }

  render() {
    switch (this.props.status) {
      case "fail":
        return <Text color={Color.Red}>Ã—</Text>;
      case "pass":
        return <Text color={Color.Green}>+</Text>;
      case "skip":
        return <Text color={Color.Yellow}>- skip</Text>;
      case "todo":
        return <Text color={Color.Magenta}>- todo</Text>;
    }
  }
}

export function testNameText(
  status: "fail" | "pass" | "skip" | "todo",
  name: string,
  indent = 0,
): ScribblerJsx.Element {
  return (
    <Line indent={indent + 1}>
      <StatusText status={status} /> <Text color={Color.Gray}>{name}</Text>
    </Line>
  );
}
