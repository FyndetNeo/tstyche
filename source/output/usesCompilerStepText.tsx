import { Path } from "#path";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";

interface ProjectNameTextProps {
  filePath: string;
}

class ProjectNameText implements ScribblerJsx.ElementClass {
  props: ProjectNameTextProps;

  constructor(props: ProjectNameTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    return (
      <Text color={Color.Gray}>
        {" with "}
        {Path.relative("", this.props.filePath)}
      </Text>
    );
  }
}

export function usesCompilerStepText(
  compilerVersion: string,
  tsconfigFilePath: string | undefined,
  options?: { prependEmptyLine: boolean },
): ScribblerJsx.Element {
  let projectPathText: ScribblerJsx.Element | undefined;

  if (tsconfigFilePath != null) {
    projectPathText = <ProjectNameText filePath={tsconfigFilePath} />;
  }

  return (
    <Text>
      {options?.prependEmptyLine === true ? <Line /> : undefined}
      <Line>
        <Text color={Color.Blue}>{"uses"}</Text>
        {" TypeScript "}
        {compilerVersion}
        {projectPathText}
      </Line>
      <Line />
    </Text>
  );
}
