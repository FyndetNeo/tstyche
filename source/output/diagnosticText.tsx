import { type Diagnostic, DiagnosticCategory } from "#diagnostic";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";
import { CodeSpanText } from "./CodeSpanText.js";

interface DiagnosticTextProps {
  diagnostic: Diagnostic;
}

class DiagnosticText implements ScribblerJsx.ElementClass {
  props: DiagnosticTextProps;

  constructor(props: DiagnosticTextProps) {
    this.props = props;
  }

  render(): ScribblerJsx.Element {
    const code =
      typeof this.props.diagnostic.code === "string" ? (
        <Text color={Color.Gray}> {this.props.diagnostic.code}</Text>
      ) : undefined;

    const text = Array.isArray(this.props.diagnostic.text) ? this.props.diagnostic.text : [this.props.diagnostic.text];

    const message = text.map((text, index) => (
      <Text>
        {index === 1 ? <Line /> : undefined}
        <Line>
          {text}
          {code}
        </Line>
      </Text>
    ));

    const related = this.props.diagnostic.related?.map((relatedDiagnostic) => (
      <DiagnosticText diagnostic={relatedDiagnostic} />
    ));

    const codeSpan = this.props.diagnostic.origin ? (
      <Text>
        <Line />
        <CodeSpanText {...this.props.diagnostic.origin} />
      </Text>
    ) : undefined;

    return (
      <Text>
        {message}
        {codeSpan}
        <Line />
        <Text indent={2}>{related}</Text>
      </Text>
    );
  }
}

export function diagnosticText(diagnostic: Diagnostic): ScribblerJsx.Element {
  let prefix: ScribblerJsx.Element | undefined;

  switch (diagnostic.category) {
    case DiagnosticCategory.Error: {
      prefix = <Text color={Color.Red}>{"Error: "}</Text>;
      break;
    }

    case DiagnosticCategory.Warning: {
      prefix = <Text color={Color.Yellow}>{"Warning: "}</Text>;
      break;
    }

    default:
      break;
  }

  return (
    <Text>
      {prefix}
      <DiagnosticText diagnostic={diagnostic} />
    </Text>
  );
}
