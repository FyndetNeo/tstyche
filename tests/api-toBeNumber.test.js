import { strict as assert } from "node:assert";
import { test } from "mocha";
import * as tstyche from "tstyche";
import { matchSnapshot } from "./__utilities__/assert.js";
import { getFixtureFileUrl, getTestFileName } from "./__utilities__/fixture.js";
import { normalizeOutput } from "./__utilities__/output.js";
import { spawnTyche } from "./__utilities__/tstyche.js";

const testFileName = getTestFileName(import.meta.url);
const fixtureUrl = getFixtureFileUrl(testFileName);

test("'toBeNumber' implementation", function() {
  assert(typeof tstyche.expect().type.toBeNumber === "function");
});

test("toBeNumber", async function() {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

  await matchSnapshot(normalizeOutput(stdout), {
    fileName: `${testFileName}-stdout`,
    testFileUrl: import.meta.url,
  });

  await matchSnapshot(stderr, {
    fileName: `${testFileName}-stderr`,
    testFileUrl: import.meta.url,
  });

  assert.equal(exitCode, 1);
});
