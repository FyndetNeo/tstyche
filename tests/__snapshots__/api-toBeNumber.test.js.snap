// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeNumber: stderr 1`] = `
"Error: Type 'number' is not identical to type 'void'.

   7 |   expect(returnsNumber()).type.toBeNumber();
   8 | 
>  9 |   expect(returnsVoid()).type.toBeNumber();
     |                              ^
  10 | });
  11 | 
  12 | test("is NOT number?", () => {

       at ./__typetests__/toBeNumber.tst.ts:9:30 ❭ is number?

Error: Type 'number' is identical to type 'number'.

  13 |   expect(returnsVoid()).type.not.toBeNumber();
  14 | 
> 15 |   expect(returnsNumber()).type.not.toBeNumber();
     |                                    ^
  16 | });
  17 | 

       at ./__typetests__/toBeNumber.tst.ts:15:36 ❭ is NOT number?

"
`;

exports[`toBeNumber: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

fail ./__typetests__/toBeNumber.tst.ts
  × is number?
  × is NOT number?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      2 failed, 2 total
Assertions: 2 failed, 2 passed, 4 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
